### PR TITLE
abstract equals and hashCode

### PR DESCRIPTION
**Pull request by elharo**
@suztomo I noticed a subtle issue here where two instances of different subclasses could theoretically compare equal. This prevents that by making equals and hashCode abstract in the superclass.

### CODE REVIEW

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/MethodSymbol.java [hunk 1]
```diff
@@ -65,18 +65,16 @@ public boolean equals(Object other) {
...
     }
-    if (!super.equals(other)) {
-      return false;
-    }
     MethodSymbol that = (MethodSymbol) other;
     return isInterfaceMethod == that.isInterfaceMethod
         && name.equals(that.name)
-        && descriptor.equals(that.descriptor);
+        && descriptor.equals(that.descriptor)
+        && this.getClassBinaryName().equals(that.getClassBinaryName());
```
Comment (suztomo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/MethodSymbol.java:72]: Checking the equality of class name here is more clear than `super.equals`. Nice.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/Symbol.java [hunk 2]
```diff
@@ -47,21 +46,10 @@ String getClassBinaryName() {
   }
 
   @Override
-  public boolean equals(Object other) {
-    if (this == other) {
-      return true;
-    }
-    if (other == null || getClass() != other.getClass()) {
-      return false;
-    }
```
Comment (suztomo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/Symbol.java:56]: > two instances of different subclasses could theoretically compare equal.

Doesn't this if statement prevent that?

Comment (elharo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/Symbol.java:56]: You're right. It does.

It still means subclasses can be created with additional data that is not considered in an equality comparison. Mostly it's a little disconcerting to have an equals method that should always be overridden, especially in an abstract class.

Comment (suztomo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/Symbol.java:56]: > It still means subclasses can be created with additional data that is not considered in an equality comparison. 

I'm not sure this change is an improvement on that. Even with this PR, a subclass can declare an additional data that's not considered in equality comparison.

**Comment by suztomo** [comment]
One benefit of this PR is equality for the classes is in within the files. Readers do not need to check superclass's equality.

**Pull request merged**